gradient_descent/gradient_descent.py

```diff
@@ -0,0 +1,121 @@
+"""
+Implementation of gradient descent algorithm for minimizing cost of a linear hypothesis function.
+"""
+import numpy
+
+# List of input, output pairs
+train_data = (((5, 2, 3), 15), ((6, 5, 9), 25),
+              ((11, 12, 13), 41), ((1, 1, 1), 8), ((11, 12, 13), 41))
+test_data = (((515, 22, 13), 555), ((61, 35, 49), 150))
+parameter_vector = [2, 4, 1, 5]
+m = len(train_data)
+LEARNING_RATE = 0.009
+
+
+def _error(example_no, data_set='train'):
+    """
+    :param data_set: train data or test data
+    :param example_no: example number whose error has to be checked
+    :return: error in example pointed by example number.
+    """
+    return calculate_hypothesis_value(example_no, data_set) - output(example_no, data_set)
+
+
+def _hypothesis_value(data_input_tuple):
+    """
+    Calculates hypothesis function value for a given input
+    :param data_input_tuple: Input tuple of a particular example
+    :return: Value of hypothesis function at that point.
+    Note that parameter input value is fixed as 1.
+    Also known as 'biased input' in ML terminology and the parameter associated with it
+    is known as 'biased parameter'.
+    """
+    hyp_val = 0
+    for i in range(len(parameter_vector) - 1):
+        hyp_val = hyp_val + data_input_tuple[i]*parameter_vector[i+1]
+    hyp_val = hyp_val + 1*parameter_vector[0]
```

Comment:
As I have mentioned in comments, there is always a biased input in Artificial Neural Networks, whose value is fixed as 1. I wanted to explicitly mention this fact in the code too, that's why I have written `1*parameter_vector[0]` instead of `parameter[0]`.

But now, I guess it's better to keep it `parameter[0]`. Thanks for pointing it out.